### ISSUE TITLE
Poco::Net::SocketConnector unregistering

### ISSUE THREAD

**Issue by Missmiaom** [pocoproject/poco#2513]
#### Expected behavior
Deadlock occurs when the SocketConnector object is destructed

#### Steps to reproduce the problem
Please allow me to post my code and stack diagram. To protect privacy, allow me to use *ip* and *port* instead of real data.

```
    class ClientServiceHandler
    {
    public:
        ClientServiceHandler(Poco::Net::StreamSocket& socket, Poco::Net::SocketReactor& reactor)
        {
            Poco::Thread::sleep(500); //Just to better reproduce the problem
        }
    };

    ...
    //test main
    while (1)
    {
        Poco::Net::SocketReactor reactor;
        auto sc = new Poco::Net::SocketConnector<ClientServiceHandler>(Poco::Net::SocketAddress(ip, port), reactor);
        
        Poco::Thread thread;
        thread.startFunc([&reactor]() {reactor.run(); });
        
        Poco::Thread::sleep(50);
        delete sc;

        poco_information(Poco::Logger::root(), "finish!");
    }
```

![7f45b67ed1457274e5ce4419baccc15](https://user-images.githubusercontent.com/10678668/47282216-f307de00-d610-11e8-9797-ba9da0a2dc11.png)

As shown in the figure above, there are two threads. The thread above executes the destructor, and the thread below is the onWriteable after the connection is successful.

Boxes of the same color compete for the same lock. A sequence number indicates the sequence in which the program runs

#### POCO version
1.9.1

#### Compiler and version
VS2017

#### Operating system and version
Windows 10

Could it be a problem with my usage? It's important to me to be grateful that someone pointed out my mistake or fixed this bug.

**Comment by aleks-f** [comment]
I spent quite a bit of time recently fixing some [threading issues in reactor](https://github.com/pocoproject/poco/issues/2345) and made sure ThreadSanitizer runs are clean. It is possible that there are some leftovers, but from briefly looking at your case, I can not see a deadlock there. But, it seems that you never stop the reactor or join the thread. Also, it is best practice for handlers to delete and unregister handling functions (if any are registered, but what's the point of service handler with no handling functions registered) themselves when done doing the work:

> [When the ServiceHandler is done, it must destroy itself.](https://github.com/pocoproject/poco/blob/poco-1.9.1/Net/include/Poco/Net/SocketConnector.h#L70)

**Comment by aleks-f** [comment]
@Missmiaom I took a bit better look at this. You create the `SocketConnector` and then immediately destroy it, while the reactor is still running. So, what happens is destructor is called in the main thread, while `onConect()` is being called in another thread - reactor should be stopped, and the thread joined, before the connector is destroyed.

But, it is better to unregister immediately in any connector handler, so I have reclassified this as an enhancement.
